src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/Serialization/DirectiveTokenDescriptorConverter.cs

```diff
@@ -0,0 +1,36 @@
+﻿using Microsoft.AspNetCore.Razor.Language;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+
+namespace Microsoft.NET.Sdk.Razor.SourceGenerators;
+
+internal sealed class DirectiveTokenDescriptorConverter : JsonConverter<DirectiveTokenDescriptor>
+{
+    public override DirectiveTokenDescriptor? ReadJson(JsonReader reader, Type objectType, DirectiveTokenDescriptor? existingValue, bool hasExistingValue, JsonSerializer serializer)
+    {
+        var obj = JObject.Load(reader);
```

Comment:
I know this is a prototype, but these serializers definitely need to be looked at for performance before putting them in the product. I'm currently rewriting the existing JSON.NET converters in the tooling layer because they do horrible inefficient things like using `JObject` within a converter.

Reply:
Yes, right now the serialization is not great. My plan was to get it to point where we can profile it, then start improving, although it sounds like you're already re-writing some of them anyway.

Reply:
Yeah, it'd be good to get us to a place where the serialization logic can just go into Shared.Utilities. TagHelperDescriptor serialization currently represents a large amount of memory allocation in Razor scenarios. So, there's already a fair amount of analysis that's been done around what's good/bad with JSON.NET. https://github.com/dotnet/razor/pull/8675 rewrites the existing JSON.NET converters in the tooling layer and avoids changing behavior, since we still have to read the project.razor.json files that are already in the wild. Once this goes in, I'll introduce versioning into serialization so that we can use a much tighter format.

See https://github.com/dotnet/razor/issues/8113 for some of the details.